rasa_core/training/structures.py

```diff
@@ -89,36 +111,150 @@ def create_copy(self, use_new_id):
+                    # form deactivated, so form succeeded,
+                    # so add story string with form prefix
+                    result += self.as_story_string_helper.form_string
+                    # remove all stored story strings
+                    self._reset_stored_strings()
+
+                result += self._bot_string(s)
+
+            elif isinstance(s, FormValidation):
+                self.as_story_string_helper.form_validation = s.validate
+
+            elif isinstance(s, ActionExecutionRejected):
+                if (s.action_name ==
+                        self.as_story_string_helper.active_form):
+                    # form rejected
+                    self.as_story_string_helper.form_rejected = True
+
+            elif isinstance(s, ActionExecuted):
+                if self._is_action_listen(s):
+                    pass
+                elif self.as_story_string_helper.active_form is None:
+                    result += self._bot_string(s)
+                else:
+                    # form is active
+                    if self.as_story_string_helper.form_rejected:
+                        if (self.as_story_string_helper.form_validation and
+                                s.action_name ==
+                                self.as_story_string_helper.active_form):
+                            result += self._bot_string(
+                                    ActionExecuted(ACTION_LISTEN_NAME))
+                            result += self.as_story_string_helper.form_string
+                        else:
+                            result += \
+                                self.as_story_string_helper.no_form_string
+                        # form rejected, add story string without form prefix
+                        result += self._bot_string(s)
+                    else:
+                        # form succeeded, so add story string with form prefix
+                        result += self.as_story_string_helper.form_string
+                        result += self._bot_string(s, FORM_PREFIX)
+
+                    # remove all stored story strings
+                    self._reset_stored_strings()
+
+                    if (s.action_name ==
+                            self.as_story_string_helper.active_form):
+                        # form was successfully executed
+                        self.as_story_string_helper.form_rejected = False
+
+                self.as_story_string_helper.form_validation = True
+
+            elif isinstance(s, SlotSet):
+                if self.as_story_string_helper.active_form is None:
+                    result += self._bot_string(s)
+                else:
+                    # form is active
+                    # it is not known whether the form will be
+                    # successfully executed, so store this
+                    # story string for later
+                    # slots should be always printed without prefix
+                    self._store_bot_strings(s)
```

Comment:
in general this whole section is one giant mess 🙈 but at the same time i'm not sure there's anything we can do to improve that

Reply:
@tmbo could you please take a look at this

Reply:
i don't think we should wait for toms review on this tbh. given we just want to release on monday. He said it was fine if it was a mess no?

Reply:
ye but he didn't review this bit. I think we can merge like this, and later when he take look, update it, if needed

Reply:
oki doki